network/p2p/peerstore/peerstore_test.go

```diff
@@ -77,3 +79,343 @@ func TestPeerstore(t *testing.T) {
 	require.Equal(t, 7, len(peers))
 
 }
+
+func testPhonebookAll(t *testing.T, set []string, ph Phonebook) {
+	actual := ph.GetAddresses(len(set), PhoneBookEntryRelayRole)
+	for _, got := range actual {
+		ok := false
+		for _, known := range set {
+			if got == known {
+				ok = true
+				break
+			}
+		}
+		if !ok {
+			t.Errorf("get returned junk %#v", got)
+		}
+	}
+	for _, known := range set {
+		ok := false
+		for _, got := range actual {
+			if got == known {
+				ok = true
+				break
+			}
+		}
+		if !ok {
+			t.Errorf("get missed %#v; actual=%#v; set=%#v", known, actual, set)
+		}
+	}
+}
+
+func testPhonebookUniform(t *testing.T, set []string, ph Phonebook, getsize int) {
+	uniformityTestLength := 250000 / len(set)
+	expected := (uniformityTestLength * getsize) / len(set)
+	counts := make(map[string]int)
+	for i := 0; i < len(set); i++ {
+		counts[set[i]] = 0
+	}
+	for i := 0; i < uniformityTestLength; i++ {
+		actual := ph.GetAddresses(getsize, PhoneBookEntryRelayRole)
+		for _, xa := range actual {
+			if _, ok := counts[xa]; ok {
+				counts[xa]++
+			}
+		}
+	}
+	min, max := math.MaxInt, 0
+	for _, count := range counts {
+		if count > max {
+			max = count
+		}
+		if count < min {
+			min = count
+		}
+	}
+	// TODO: what's a good probability-theoretic threshold for good enough?
+	if max-min > (expected / 5) {
+		t.Errorf("counts %#v", counts)
+	}
+}
+
+func TestArrayPhonebookAll(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	set := []string{"a:4041", "b:4042", "c:4043", "d:4044", "e:4045", "f:4046", "g:4047", "h:4048", "i:4049", "j:4010"}
+	ph, err := MakePhonebook(1, 1*time.Millisecond)
+	require.NoError(t, err)
+	for _, addr := range set {
+		entry := makePhonebookEntryData("", PhoneBookEntryRelayRole, false)
+		info, _ := PeerInfoFromDomainPort(addr)
+		ph.AddAddrs(info.ID, info.Addrs, libp2p.AddressTTL)
+		ph.Put(info.ID, "addressData", entry)
+	}
+	testPhonebookAll(t, set, ph)
+}
+
+func TestArrayPhonebookUniform1(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	set := []string{"a:4041", "b:4042", "c:4043", "d:4044", "e:4045", "f:4046", "g:4047", "h:4048", "i:4049", "j:4010"}
+	ph, err := MakePhonebook(1, 1*time.Millisecond)
+	require.NoError(t, err)
+	for _, addr := range set {
+		entry := makePhonebookEntryData("", PhoneBookEntryRelayRole, false)
+		info, _ := PeerInfoFromDomainPort(addr)
+		ph.AddAddrs(info.ID, info.Addrs, libp2p.AddressTTL)
+		ph.Put(info.ID, "addressData", entry)
```

Comment:
nit: I like replacing these strings with `addressDataKey` but you missed them in tests.